Fix adding orphaned user to org

fosterseth opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Take all the roles of orphan User A, find all ancestors to these roles. The org admin must have 1 or more of these ancestor roles. If not, the admin cannot add User A to the org.

The conditional was incorrectly checking that scenario -- it wouldn't add the User A, if that User A had a role with an ancestor that the org admin also had.

This change fixes that.

I modified the functional test for adding an orphan. Now the `rando` user has membership to an organization credential, but `rando` is not a member of the org. Now when org admin tries to add `rando`, it will pass, where before this assertion would have failed. The test passed previously because rando had no roles at all, so the `not  .. .exists()` conditional was always met. 
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 10.0.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

AlanCoding commented:
> Take all the roles of orphan User A, find all ancestors to these roles. The org admin must have all of these ancestor roles.

That wording isn't quite right. The org admin must have 1 or more ancestor of each role, not all ancestors.

fosterseth commented:
> That wording isn't quite right. The org admin must have 1 or more ancestor of each role, not all ancestors.

@AlanCoding  you're right, I changed the wording

fosterseth commented:
@AlanCoding 

however, this test is now failing with the change,

https://github.com/ansible/awx/blob/48414f6dab13527395d3918791120d63a985f9a3/awx/main/tests/functional/test_rbac_role.py#L56

`nonmember` has one role, the inventory admin role, which has an ancestor of org admin. Therefore, the org admin _can_ add this nonmember to the organization (fulfills the iff rule for orphans)

We need an inventory (or some other resource) that isn't part of the organization, then these assertions should fail. 

I can add these lines
```
inventory.organization = None
inventory.save()
```

AlanCoding commented:
huh, I think I would prefer to keep the inventory admin permission in place in that case.

I think it would be better to distinguish between non-members and orphans. Orphans are not members of any organization - and this was only supposed to be a transient state to make management easier.

So I'm curious if you add `nonmember` in that test to the member_role of some other arbitrary org, it'll make the assertions line up again.

fosterseth committed:
Allow orphaned user to be added to org

Fixed bug where an org admin was not able to add
an orphaned user to the org, in the case where the
orphan had an ancestor role that matched one of the
roles for of the org admin.

scenario to fix -- sue is member of cred1, where cred1 is
part of org1. org1 admin cannot add sue to org1, because
the cred1 role for sue has an ancestor to org1 role. The org1
admin cannot change or attach sue to org1.

tower issue #4198 and #4197

fosterseth commented:
I agree, since we have other tests for adding orphans, what this specific test should be doing is determining if a member of another org can be added (it cannot) 

I changed the test by adding these lines

```python
other_org = Organization.objects.create(name="other_org")
other_org.member_role.members.add(nonmember)
```

the assertions work

ryanpetrello commented:
regate
